packages/gatsby/content/features/constraints.md

```diff
@@ -23,11 +23,17 @@ Going back to the constraint engine, the *facts* are the definitions created by
 
 The following predicates provide information about the current state of your project and are meant to be used in the dependencies of your own rules (check the recipes for examples how to use them in practice). Note that the `/<number>` syntax listed at the end simply is the predicate arity (number of arguments it takes).
 
+The notation on this page uses `-`, `+` and `?` as prefix for the predicate parameters. These values are used commonly in prolog documentation and mean
+
+- `+`: this value is considered input and must be instantiated
+- `-`: this value is considered output and will be instantiated by the predicate, though you can provide a value to verify that the value matches the predicate
+- `?`: this value can be instantiated or not, both will work
```

Comment:
I'm not convinced this section makes things any simpler:

- I myself don't entirely see the difference between `?WorkspaceCwd` and `+WorkspaceCwd`
- the meaning of "instantiated" in this context is I think a bit hard to grasp here
- it prevents to copy-paste the predicates and see their effects

Maybe it would be better with real-life example but it might be out of scope for this documentation 🤔